### ISSUE TITLE
untangle key location from encryption/authentication type byte

### ISSUE THREAD

**Issue by ThomasWaldmann** [borgbackup/borg#2153]
As already seen in #510, changing the key storage location (repokey vs. keyfile) is non-trivial because the type byte in the chunks does not only tell the enc/auth type, but also the key type (== where the key is located).

Maybe we could just say "it doesn't matter where the key is located as long as we have the right one" and just look at the well-known places (== in the repo config, locally in the keys dir).

If we can do it like that, we could map the keyfile type byte and the repokey type byte to the same processing logic. Changing the key location then would really be that: move it from repo config to keys dir (or vice versa).

**Comment by textshell** [comment]
Just looking for the key would be a possibility yes. The other way i see is that the type byte is still used on the manifest, but ignored in all other code paths.
It could be nice to have an indication where the key is supposed to be. Restoring a key backup also might need to know where to place it. Of course it could also ask or require an option.

**Comment by enkore** [comment]
If we just use a fixed lookup order (eg. first local key file then repokey) then we can run into issues with multiple clients, eg. going from key file to repokey would necessitate manually deleting all key files on all clients, otherwise things like change-passphrase would only apply to those clients who don't have the key file anymore.

So still using it for the manifest makes sense, and just ignoring it - in general - otherwise. Ie. key.py wouldn't care apart from the key-identification logic, which only looks at the manifest.

If we lose the manifest (ie. we are borg check) we can look at some other chunks and/or use a fixed lookup order.

**Comment by ThomasWaldmann** [comment]
If we treat the manifest chunk differently (by looking at its type byte to select the key), it makes the code pathes taken with/without a manifest different. So maybe it is easier and less error prone not doing that.

We have to be careful btw when doing the key lookup - the first **working** key counts, not the first **discovered** key. Doing it that way, we could first try the repokey (avoiding the issue mentioned above by @enkore) and then the local key. An attacker could write an invalid repokey into the config, but if it does not work, we would just try the (correct) local key.

**Comment by textshell** [comment]
The code pathes are already separate.
Detection of the key type is in `key_factory` that does (plus in KeyManager, maybe other places):
```
def key_factory(repository, manifest_data):
    key_type = manifest_data[0]
    if key_type == KeyfileKey.TYPE:
...
```
while the code that needs to be remove is in the first line of every `decrypt` method:
```
    def decrypt(self, id, data):
        if data[0] != self.TYPE:
            id_str = bin_to_hex(id) if id is not None else '(unknown)'
            raise IntegrityError('Chunk %s: Invalid encryption envelope' % id_str)
```

**Comment by ThomasWaldmann** [comment]
Fixed by #6415. It is using the manifest key-type byte to determine where the key should be.
